File: core/vehicle.go
package core

import (
	"errors"
	"github.com/louisevanderlith/husk/hsk"
	"github.com/louisevanderlith/husk/op"
	"github.com/louisevanderlith/husk/records"
	"github.com/louisevanderlith/husk/validation"
	"github.com/louisevanderlith/vehicle/core/bodytype"
	"time"
)

type Vehicle struct {
	VINKey    hsk.Key
	FullVIN   string
	Series    SeriesInfo
	Colour    string
	PaintNo   string
	Engine    Engine
	Gearbox   Gearbox
	BodyStyle bodytype.Enum
	Doors     int
	Extra     []string
	Spare     bool
	Service   bool
	Condition string
	Issues    string
	Mileage   int

	Info          string `hsk:"size(128)"`
	Year          int    `orm:"null"`
	HasNatis      bool   `hsk:"default(false)"`
	EstValue      int64
	LicenseExpiry time.Time
}

func (o Vehicle) Valid() error {
	err := validation.Struct(o)
	if err != nil {
		return err
	}

	if o.Year > 0 && o.Year > time.Now().Year() {
		errors.New("year can't be in the future")
	}

	if o.Mileage < 0 {
		return errors.New("mileage can't be negative")
	}

	if o.HasNatis && o.LicenseExpiry.Before(time.Now()) {
		return errors.New("license has already expired")
	}

	return nil
}

func GetVehicles(page, pagesize int) (records.Page, error) {
	return ctx.Vehicles.Find(page, pagesize, op.Everything())
}

func GetVehicle(key hsk.Key) (hsk.Record, error) {
	return ctx.Vehicles.FindByKey(key)
}

func (obj Vehicle) Create() (hsk.Key, error) {
	if ctx.Vehicles.Exists(byFullVIN(obj.FullVIN)) {
		return nil, errors.New("vehicle VIN already exists")
	}

	return ctx.Vehicles.Create(obj)
}

File: lib/bodies/vehicle.dart
import 'package:mango_ui/keys.dart';

import 'engine.dart';
import 'gearbox.dart';
import 'series.dart';

class Vehicle {
  final Key vinKey;
  final String fullVin;
  final Series series;
  final String colour;
  final String paintNo;
  final Engine engine;
  final Gearbox gearbox;
  final String bodyStyle;
  final num doors;
  final List<String> extra;
  final bool spare;
  final bool service;
  final String condition;
  final String issues;
  final num mileage;

  Vehicle(
      this.vinKey,
      this.fullVin,
      this.series,
      this.colour,
      this.paintNo,
      this.engine,
      this.gearbox,
      this.bodyStyle,
      this.doors,
      this.extra,
      this.spare,
      this.service,
      this.condition,
      this.issues,
      this.mileage);

  Map<String, dynamic> toJson() {
    return {
      "VINKey": this.vinKey,
      "FullVIN": this.fullVin,
      "Series": this.series,
      "Colour": this.colour,
      "PaintNo": this.paintNo,
      "Engine": this.engine,
      "Gearbox": this.gearbox,
      "BodyStyle": this.bodyStyle,
      "Doors": this.doors,
      "Extra": this.extra,
      "Spare": this.spare,
      "Service": this.service,
      "Condition": this.condition,
      "Issues": this.issues,
      "Mileage": this.mileage
    };
  }
}
